internal/pkg/gateway/commit/notifier.go

```diff
@@ -0,0 +1,91 @@
+/*
+Copyright 2021 IBM All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package commit
+
+import (
+	commonProto "github.com/hyperledger/fabric-protos-go/common"
+	ordererProto "github.com/hyperledger/fabric-protos-go/orderer"
+	peerProto "github.com/hyperledger/fabric-protos-go/peer"
+	"github.com/hyperledger/fabric/common/ledger/blockledger"
+)
+
+// Notifier enables notification of commits to a peer ledger.
+type Notifier struct {
+	channelFactory ChannelFactory
+}
+
+// NewNotifier returns a new commit notifier that obtains channels from a given channel factory.
+func NewNotifier(channelFactory ChannelFactory) *Notifier {
+	return &Notifier{
+		channelFactory: channelFactory,
+	}
+}
+
+// Notify the supplied channel when the specified transaction is committed to a channel's ledger.
+func (notifier *Notifier) Notify(commit chan<- peerProto.TxValidationCode, channelID string, transactionID string) error {
+	channel, err := notifier.channelFactory.Channel(channelID)
+	if err != nil {
+		return err
+	}
+
+	blockReader := channel.Reader()
+	go readCommit(commit, blockReader, transactionID)
+
+	return nil
+}
+
+func readCommit(commit chan<- peerProto.TxValidationCode, blockReader blockledger.Reader, transactionID string) {
+	blockIter, _ := blockReader.Iterator(seekNextCommit())
```

Comment:
- The code is not behaving as you describe. The iterator is obtained once and used to obtain blocks until the required transaction is located
- I have no intention of exposing this function directly via gRPC. It is intended to be used internally by the Gateway to identify when transactions submitted to the orderer have committed on the local peer
- Some kind of timeout will be required once the end-to-end behaviour is implemented, and perhaps this code might want a `cancel` channel so it can be signalled to stop processing when that timeout is hit. Agile development though. Small steps at a time. I don't want to boil the ocean in one PR

Reply:
Yes, absolutely that would be a great thing to do in the future

Reply:
As this impacts the peer's performance indirectly, I am not sure what motivates to push this to the future and not getting it right in the first cut.

Reply:
Before making performance optimizations I would normally want to have benchmarking data to:

1. Confirm there is a performance issue worth investing effort, and adding complexity and risk to fix
2. Ensure the optimizations actually deliver results

Since this code is not yet in Fabric, and this PR doesn't deliver any code that is actually executed at runtime, I would appreciate your guidance on how to obtain this data.

It sounds like the the performance hot spot is getting and using blockledger iterators, so we should avoid doing that more than necessary. Is it worth exploring the possibility of implementing optimizations within the iterator itself so all consuming code can benefit? Again, I would appreciate your collaboration on how implementing optimizations based on measured performance.